### DISCUSSION TITLE
New Pep 654 (`ExceptionGroup`)

### DISCUSSION THREAD

**Discussion by ghost** [pydantic/pydantic#2413] [General]
A couple of days ago a new [Pep](https://www.python.org/dev/peps/pep-0654/) has been put forward that aims to introduce a language construct for wrapping multiple errors that occurred together. As main motivation the pep uses async frameworks such as asyncio and trio but also the hypothesis testing library. I think features from this pep might be useful for pydantic's `ValidationError` too? Maybe we could get a nice syntax  (possibly together with the pattern matching peps) for retrieving and handling individual errors from a `ValidationError`?

**Comment by ghost** [comment]
there is also a [thread on the mailing list](https://mail.python.org/archives/list/python-dev@python.org/thread/L5Q27DVKOKZCDNCAWRIQVOZ5DZCZHLRM/) about the api and its usability

**Comment by samuelcolvin** [comment]
It's a really interesting idea and definitely shadows closely how pydantic things about a validation error - one or more exceptions when validating data.

My reservations are:
* Can it be back ported? If so, (provided it's performant) we could definitely adopt it. If not, it'll be a long time before pydantic drops support for python that doesn't support this, even if it gets accepted asap.
* So what? 99% of code does one of the following, and none of these cases really needs exception groups:
  *  lets validation errors propagate and be displayed using `__str__` in the terminal,
  * returns `error.json()` to a request, or
  * does it's own processing of `error.errors()`

In the end, it looks a bit like a language feature for language geeks, without much practical applicability, see [this tweet](https://twitter.com/mitsuhiko/status/1364166383714635779). But then again, perhaps many feature start out that way, and mere mortals like me only see the point when they're in widespread use.

**Reply by ghost** [reply]
What I'd like to get out from `ExceptionGroup` both for usage in async code and libraries like pydantic and hypothesis would be a standardized way for picking apart these things. Currently in pydantic I mostly do your "return error.json() to a request" because "own processing of error.errors()" is to messy/ to much effort just to make the error messages a bit nicer for the requester.

**Reply by ghost** [reply]
Maybe for pydantic this is somewhat of a chicken/egg problem as currently there are only three different exceptions one can raise `ValueError`, `TypeError` and `AssertionError` (leaving out custom exceptions which I for one have not yet used) so there is no big need of dispatching based on individual exceptions. A somewhat over-engineered toy example how stuff could look with the new syntax:

```python
### some custom exceptions
class TotallyWrongRecipientError(ValueError): pass

class MessageForTheNeighborsError(ValueError): pass

class MessageIsSpamError(ValueError): pass


Us = TypeVar('Us')
Neighbors = TypeVar('Neighbors')

class Postcard(GenericModel, Generic[Us, Neighbors]):
    recipient: str
    content: str
        
    @validator('recipient')
    def check_if_message_is_for_me(cls, value):
        if value in Us.__args__:   # this doesn't work btw, is this something that will work in pydantic 1.8?
            return value
        elif value in Neighbors.__args__:
            raise MessageForTheNeighborsError(f'{value} is for the neighbors again')
        else:
            raise TotallyWrongRecipientError(f'{value} is not for anyone we know')
        
    @validator('content')
    def check_if_message_contains_spam(cls, value):
        if any(spam in value for spam in ['foo', 'bar', 'hello world']):
            raise MessageIsSpamError(f'"{value}" contains spam')
        else:
            return value
            
### Usage
us = Literal['Alice', 'Sascha']        
neighbors = Literal['Bob', 'Charlie']

def sort_message(msg, fridge, trash, box):
    try:
        fridge.append(Postcard[us, neighbors](**msg))
    except *TotallyWrongRecipientError:
        trash.append(msg)
    except *MessageIsSpamError:
        trash.append(msg)
    except *MessageForTheNeighborsError:
        box.append(msg)

```
